runtime/auth/aws-signing-common/common/src/aws/smithy/kotlin/runtime/auth/awssigning/AbstractAwsChunkedByteReadChannel.kt

```diff
@@ -0,0 +1,204 @@
+        if (length == 0 || !ensureValidChunk()) {
+            return 0
+        }
+
+        var bytesWritten = 0
+
+        while (bytesWritten != length) {
+            val numBytesToWrite = minOf(length, chunk!!.size - chunkOffset)
+
+            chunk!!.copyInto(sink, offset + bytesWritten, chunkOffset, chunkOffset + numBytesToWrite)
+
+            bytesWritten += numBytesToWrite
+            chunkOffset += numBytesToWrite
+
+            // if we've exhausted the current chunk, exit without suspending for a new one
+            if (chunkOffset >= chunk!!.size) { break }
+        }
+
+        return bytesWritten
+    }
+
+    /**
+     * Ensures that the internal [chunk] is valid for reading. If it's not valid, try to load the next chunk. Note that
+     * this function will suspend until the whole chunk has been loaded.
+     *
+     * @return true if the [chunk] is valid for reading, false if it's invalid (chunk data is exhausted)
+     */
+    suspend fun ensureValidChunk(): Boolean {
+        // check if the current chunk is still valid
+        if (chunk != null && chunkOffset < chunk!!.size) { return true }
+
+        // if not, try to fetch a new chunk
+        val nextChunk = if (chan.isClosedForRead && hasLastChunkBeenSent) {
+            null
+        } else if (chan.isClosedForRead && !hasLastChunkBeenSent) {
+            hasLastChunkBeenSent = true
+            getChunk(byteArrayOf()) + if (!trailingHeaders.isEmpty()) { getTrailingHeadersChunk(trailingHeaders) } else byteArrayOf()
+        } else {
+            getChunk()
+        }
+
+        chunkOffset = 0
+        chunk = nextChunk?.plus("\r\n".encodeToByteArray()) // terminating CRLF to signal end of chunk
+        return (chunk != null)
+    }
+
+    /**
+     * Get an aws-chunked encoding of [data].
+     * If [data] is not set, read the next chunk from [chan] and add hex-formatted chunk size and chunk signature to the front.
+     * Note that this function will suspend until the whole chunk has been read.
+     * The chunk structure is: `string(IntHexBase(chunk-size)) + ";chunk-signature=" + signature + \r\n + chunk-data + \r\n`
+     *
+     * @param data the ByteArray of data which will be encoded to aws-chunked. if not provided, will default to
+     * reading up to [CHUNK_SIZE_BYTES] from [chan].
+     * @return a ByteArray containing the chunked data
+     */
+    private suspend fun getChunk(data: ByteArray? = null): ByteArray {
+        val chunkBody = data ?: chan.readRemaining(CHUNK_SIZE_BYTES)
+
+        val chunkSignature = signer.signChunk(chunkBody, previousSignature, signingConfig).signature
+        previousSignature = chunkSignature
+
+        val chunkHeader = buildString {
+            append(chunkBody.size.toString(16))
+            append(";")
+            append("chunk-signature=")
+            append(chunkSignature.decodeToString())
+            append("\r\n")
+        }.encodeToByteArray()
+
+        return chunkHeader + chunkBody
```

Comment:
Comment/no-op: This is fine for now because we are going to be updating this code soon with a refactor of `SdkByteReadChannel` but in general we want to be very cognizant of performance characteristics we use/introduce. Each `plus` operator on a `ByteArray` is an implicit copy behind the scenes. One optimization I would probably consider here is reading to a buffer with large enough capacity to handle the header + body and setting a limit for the chunk. This would avoid excessive copying and read directly into the underlying buffer. 

I call this out not to fix in this PR (again because we have some upcoming expected churn anyway) but as something to pay attention to in general and be aware of.